lib/RouteStore.js

```diff
@@ -0,0 +1,108 @@
+var createStore = require('fluxible/utils/createStore');
+var Router = require('routr');
+var queryString = require('query-string');
+var searchPattern = /\?([^\#]*)/;
+var util = require('util');
+var objectAssign = require('object-assign');
+var Immutable = require('immutable');
+
+var RouteStore = createStore({
+    storeName: 'RouteStore',
+    handlers: {
+        'NAVIGATE_START': 'handleNavigateStart',
+        'RECEIVE_ROUTES': 'handleReceiveRoutes'
+    },
+    initialize: function () {
+        this.routes = null;
+        this.currentRoute = null;
+        this.router = null;
+    },
+    handleNavigateStart: function (payload) {
+        var options = {
+            navigate: payload,
+            method: payload.method
+        };
+
+        this.currentUrl = payload.url;
+        var matchedRoute = this._matchRoute(payload.url, options);
+        if (!matchedRoute || !matchedRoute.equals(this.currentRoute)) {
+            this.currentRoute = matchedRoute;
+            this.emitChange();
+        }
+    },
+    _matchRoute: function (url, options) {
+        var self = this;
+        var route = self.getRouter().getRoute(url, options);
+        if (!route) {
+            return null;
+        }
+
+        var matchedUrl = route.url;
+        route = Immutable.fromJS(route.config).withMutations(function (r) {
+            r.set('params', Immutable.fromJS(route.params));
+            r.set('query', Immutable.fromJS(self._parseQueryString(matchedUrl)));
+        });
+
+        return route;
+    },
+    _parseQueryString: function (url) {
+        var matches = url.match(searchPattern);
+        var search;
+        if (matches) {
+            search = matches[1];
+        }
+        return (search && queryString.parse(search)) || {};
+    },
+    handleReceiveRoutes: function (payload) {
+        this.routes = objectAssign(this.routes || {}, payload);
```

Comment:
I found that this actually just made things more complicated and added unnecessary overhead. I could only see this being useful if routr handled immutable route config or if one would expect the routes to change a lot throughout the store's lifetime.